forge/routes/api/deviceLive.js

```diff
@@ -35,8 +35,44 @@ module.exports = async function (app) {
      * If the snapshot doesn't match the target, it will get a 409 (conflict)
      */
     app.post('/state', async (request, reply) => {
+        const isProvisioning = request.device.isApplicationOwned && request.body.provisioning?.deviceConfig && request.body.state === 'provisioning'
         await app.db.controllers.Device.updateState(request.device, request.body)
         if (request.device.isApplicationOwned) {
+            const device = request.device
+            if (isProvisioning) {
+                // If the device is in provisioning mode, set the target snapshot to the init snapshot
+                // generate a new snapshot for this device
+                const user = request.session?.User || { id: null }
+                try {
+                    const provisionedSnapshotOptions = {
+                        name: request.body.provisioning.name || 'Imported snapshot',
+                        description: request.body.provisioning.description || 'Snapshot imported during provisioning at ' + new Date().toISOString(),
+                        setAsTarget: true,
+                        credentialSecret: request.body.provisioning.credentialSecret || undefined,
+                        deviceConfig: request.body.provisioning.deviceConfig
+                    }
+                    const snapShot = await app.db.controllers.ProjectSnapshot.createDeviceSnapshot(
+                        device.Application,
+                        device,
+                        { id: null }, // state is transmitted from the device, so we don't have a user id
+                        provisionedSnapshotOptions
+                    )
+                    device.set('targetSnapshotId', snapShot.id)
+                    await device.save()
+                    // deliberate 6s delay for effect
+                    await new Promise(resolve => setTimeout(resolve, 6000))
```

Comment:
What effect is this for? 6 seconds feels like a lifetime when you're waiting for something happen on the device. Is it necessary or can it be shortened?

Reply:
ah, that was a testing/simulating realworld slow reply! (slipped through)